### PR TITLE
feat!: merge matchPaths and matchFiles into matchFileNames

### LINKED ISSUE #22395: Combine matchFiles and matchPaths
### Describe the proposed change(s).

Once partial string matching is removed from `matchPaths` (see #22394) we should combine `matchFiles` and `matchPaths` into one - probably in v36 too. It seems that a plain file name like `matchFiles` uses is a valid glob anyway, so the only difference is that `matchFiles` matches against both package file and lock file while `matchPaths` currently matches only against package files.

We could either combine them into `matchPaths`, or we could make a new field like `matchFileNames`.

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Merges `matchFiles` and `matchPaths` into `matchFileNames`.

## Context

Closes #22395

Declared as breaking because if a user has both `matchFiles` and `matchPaths` then whichever is placed second takes precedence.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/types.ts [hunk 1]
```diff
@@ -459,6 +458,7 @@ export type RenovateOptions =
 export interface PackageRuleInputConfig extends Record<string, unknown> {
   versioning?: string;
   packageFile?: string;
+  lockFiles?: string[];
```
Comment (viceice) [inline lib/config/types.ts:461]: unrelated change?

File: lib/util/package-rules/files.ts [hunk 2]
```diff
@@ -1,23 +1,27 @@
...
       return false;
     }
 
-    return matchFiles.some(
-      (fileName) =>
-        packageFile === fileName ||
-        (is.array(lockFiles) && lockFiles?.includes(fileName))
+    return matchFileNames.some(
+      (matchFileName) =>
+        minimatch(packageFile, matchFileName, { dot: true }) ||
```
Comment (viceice) [inline lib/util/package-rules/files.ts:20]: should we create a Minimatch object outside the loop, so Minimatch doesn't need to parse and create the regex on every iteration?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -1862,21 +1862,21 @@ For example, if you have an `examples` directory and you want all updates to tho
...
-      "matchPaths": ["examples/**"],
+      "matchFileNames": ["examples/**"],
       "extends": [":semanticCommitTypeAll(chore)"]
     }
   ]
 }
 ```
 
-If you wish to limit Renovate to apply configuration rules to certain files in the root repository directory, you have to use `matchPaths` with a `minimatch` pattern or use [`matchFiles`](#matchfiles) with an exact match.
+If you wish to limit Renovate to apply configuration rules to certain files in the root repository directory, you have to use `matchFileNames` with a `minimatch` pattern (which can include an exact file name match).
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1872]: Line 1858 mentions:

> Path rules are convenient to use if you wish to apply configuration rules to certain package files using patterns.

I think you should search for `path` as well in the `*.md` files, you'll probably find more places where `path` should be replaced with `matchFileNames` or a similar word.

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -1862,21 +1862,21 @@ For example, if you have an `examples` directory and you want all updates to tho
 {
   "packageRules": [
     {
-      "matchPaths": ["examples/**"],
+      "matchFileNames": ["examples/**"],
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1865]: Line 1858 in this file says:

> Path rules are convenient to use if you wish to apply configuration rules to certain package files using patterns.

I think this line needs to be changed as well?

You should probably search for the string `path` in the `*.md` files, to make sure you've changed all the relevant docs.

File: docs/usage/configuration-options.md [hunk 5]
```diff
@@ -2171,23 +2171,50 @@ Use the syntax `!/ /` like this:
...
 ```
 
-### matchFiles
+### matchFileNames
 
-Renovate will compare `matchFiles` for an exact match against the dependency's package file or lock file.
+Renovate will compare `matchFileNames` glob matching against the dependency's package file or lock file.
 
-For example the following would match `package.json` but not `package/frontend/package.json`:
+The following example would match `package.json` but not `package/frontend/package.json`:
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2178]:
```suggestion
The following example matches `package.json` but _not_ `package/frontend/package.json`:
```

File: lib/modules/manager/npm/extract/index.ts [hunk 6]
```diff
@@ -64,7 +64,7 @@ export async function extractPackageFile(
     const error = new Error(CONFIG_VALIDATION);
     error.validationSource = fileName;
     error.validationError =
-      'Nested package.json must not contain renovate configuration. Please use `packageRules` with `matchPaths` in your main config instead.';
+      'Nested package.json must not contain renovate configuration. Please use `packageRules` with `matchFileNames` in your main config instead.';
```
Comment (HonkingGoose) [inline lib/modules/manager/npm/extract/index.ts:67]:
```suggestion
      'Nested package.json must not contain Renovate configuration. Please use `packageRules` with `matchFileNames` in your main config instead.';
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/types.ts (hunk 1 above)
Comment (rarkins) [reply lib/config/types.ts:461]: It was missing before, and needed now because of type enforcement of `minimatch()`

File: lib/util/package-rules/files.ts (hunk 2 above)
Comment (rarkins) [reply lib/util/package-rules/files.ts:20]: Not worth it because it should always be <=3 total files (packageFile + lockFile)

**Review by HonkingGoose — APPROVED** [review]

**Review by secustor — APPROVED** [review]
